### PR TITLE
Add `parse_ints` config in json parser to support parsing int or float properly

### LINKED ISSUE #33696: json_parser: number is always converted to float64
### Component(s)

pkg/stanza

## Description

By default both the standard json library and golang jsoniter library converts number value to float64, there is an option `UseNumber` from the library, with it enabled, the value will be `json.Number` and then v.Int64() or v.Float64() returns value in corresponding type or any parse error. It would make sense to honor the original data type from the json parser perspective.

## Steps to Reproduce

use json parser to parse a sample json file. f.e. given following log line

```
{"int": 1, "float": 1.0}
```

## Expected Result

int: int64(1), float: float64(1.0)

## Actual Result

int: float64(1), float: float64(1.0)

### Collector version

latest

## Environment
OS: (e.g., "Ubuntu 20.04")
Compiler(if manually compiled): (e.g., "go 14.2")

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by newly12**
**Description:** <Describe what has changed.>

expose json iterator config in json parser

**Link to tracking Issue:** <Issue number if applicable>

Fixes #33696

**Testing:** <Describe what testing was performed and which tests were added.>

added.

**Documentation:** <Describe the documentation added.>

updated

### CODE REVIEW

File: pkg/stanza/operator/parser/json/config.go [hunk 1]
```diff
@@ -32,6 +32,38 @@ func NewConfigWithID(operatorID string) *Config {
 // Config is the configuration of a JSON parser operator.
 type Config struct {
 	helper.ParserConfig `mapstructure:",squash"`
+
+	*JsoniterConfig `mapstructure:"jsoniter_config,omitempty"`
+}
+
+type JsoniterConfig struct {
+	IndentionStep                 int    `mapstructure:"indention_step"`
```
Comment (ChrsMark) [inline pkg/stanza/operator/parser/json/config.go:40]: Do we want to expose all these options to the users?

File: pkg/stanza/operator/parser/json/parser.go [hunk 2]
```diff
@@ -36,5 +38,47 @@ func (p *Parser) parse(value any) (any, error) {
 	default:
 		return nil, fmt.Errorf("type %T cannot be parsed as JSON", value)
 	}
+
+	if p.useNumber {
```
Comment (ChrsMark) [inline pkg/stanza/operator/parser/json/parser.go:42]: Doing the conversion explicitly here makes me wonder what is the actual reason for defining the `UseNumber` setting in the `json` object's config at first place 🤔 ?

File: pkg/stanza/docs/operators/json_parser.md [hunk 3]
```diff
@@ -4,21 +4,43 @@ The `json_parser` operator parses the string-type field selected by `parse_from`
...
 
 ### Embedded Operations
 
 The `json_parser` can be configured to embed certain operations such as timestamp and severity parsing. For more information, see [complex parsers](../types/parsers.md#complex-parsers).
 
+### `jsoniter_config` Configuration
+
+The `json_parser` operator uses the [json-iterator](https://github.com/json-iterator/go) as the underlying json parser, the default config is [ConfigFastest](https://pkg.go.dev/github.com/json-iterator/go#pkg-variables).
+
+In additional, this `jsoniter_config` block allows you to configure the json parser with a custom configuration. Here are available fields that map to the corresponding fields in [json-iterator Config](https://pkg.go.dev/github.com/json-iterator/go#Config):
```
Comment (ChrsMark) [inline pkg/stanza/docs/operators/json_parser.md:27]:
```suggestion
In addition, this `jsoniter_config` block allows you to configure the json parser with a custom configuration. Here are available fields that map to the corresponding fields in [json-iterator Config](https://pkg.go.dev/github.com/json-iterator/go#Config):
```

File: pkg/stanza/operator/parser/json/config.go [hunk 4]
```diff
@@ -32,6 +32,38 @@ func NewConfigWithID(operatorID string) *Config {
...
+	MarshalFloatWith6Digits       bool   `mapstructure:"marshal_float_with_6_digits"`
+	EscapeHTML                    bool   `mapstructure:"escape_html"`
+	SortMapKeys                   bool   `mapstructure:"sort_map_keys"`
+	UseNumber                     bool   `mapstructure:"use_number"`
+	DisallowUnknownFields         bool   `mapstructure:"disallow_unknown_fields"`
+	TagKey                        string `mapstructure:"tag_key"`
+	OnlyTaggedField               bool   `mapstructure:"only_tagged_field"`
+	ValidateJsonRawMessage        bool   `mapstructure:"validate_json_raw_message"`
```
Comment (crobert-1) [inline pkg/stanza/operator/parser/json/config.go:48]: Lint is failing on this line:
```
Error: stanza/operator/parser/json/config.go:48:2: var-naming: struct field ValidateJsonRawMessage should be ValidateJSONRawMessage (revive)
	ValidateJsonRawMessage        bool   `mapstructure:"validate_json_raw_message"`
```

File: pkg/stanza/operator/parser/json/config.go (hunk 1 above)
Comment (newly12) [reply pkg/stanza/operator/parser/json/config.go:40]: I agree, my second thought on this is at least at this moment we should only expose options we needed, to not bind to the json iterator so tight.

File: pkg/stanza/operator/parser/json/parser.go (hunk 2 above)
Comment (newly12) [reply pkg/stanza/operator/parser/json/parser.go:42]: I am not sure either. Looking at the [encoding/json](https://pkg.go.dev/encoding/json#Decoder.UseNumber) it looks like not as these many options as json iterator, which makes more sense to me to only expose `use_number` option for now.

File: pkg/stanza/operator/parser/json/config.go (hunk 1 above)
Comment (newly12) [reply pkg/stanza/operator/parser/json/config.go:40]: currently `json_parser` uses `jsoniter.ConfigFastest` which has 2 more options enabled

```go
var ConfigFastest = [Config](https://pkg.go.dev/github.com/json-iterator/go#Config){
	EscapeHTML:                    [false](https://pkg.go.dev/builtin#false),
	MarshalFloatWith6Digits:       [true](https://pkg.go.dev/builtin#true),
	ObjectFieldMustBeSimpleString: [true](https://pkg.go.dev/builtin#true),
}.Froze()
```

to add the `use_number` config, it looks like we have to create a new `jsoniter.Config` object probably to have the same config so users are not surprised, I think it makes sense to allow user to be able to turn off `MarshalFloatWith6Digits` as well. any thoughts or comments on this?

Comment (djaglowski) [reply pkg/stanza/operator/parser/json/config.go:40]: I agree we should only expose what is needed, and we should consider that perhaps some day we need to switch json parsing libraries.

File: pkg/stanza/operator/parser/json/parser.go (hunk 2 above)
Comment (ChrsMark) [reply pkg/stanza/operator/parser/json/parser.go:42]: So what's the point of enabling the `UseNumber` of the `jsoniter.Config`? Is this required for some reason? If so I suggest we document this at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33699/files#diff-158189e84f05b177451492225bc83c2c23fa140cffdd4c5f9bf7db6ada8edc3aR54.

Comment (newly12) [reply pkg/stanza/operator/parser/json/parser.go:42]: Added comment, please also check this test that when UseNumber is false, regardless the data is int or float, they will be parsed as `float64`. https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33699/files#diff-f811622a59595f3ded728a9213b5e08d0c5fbe91c4cfff23b25d037f84e1f953R142-R155

File: pkg/stanza/operator/parser/json/config.go (hunk 1 above)
Comment (newly12) [reply pkg/stanza/operator/parser/json/config.go]: I was wrong about `MarshalFloatWith6Digits`, it doesn't affect unmarshal, only marshal, which is not related to the parser. so for now I only added config for `UseNumber`.

**Comment by newly12** [comment]
@djaglowski @ChrsMark could you please review again?

**Review by ChrsMark — APPROVED** [review]
It seems that are switching the json library already: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33785

**Review by ChrsMark — CHANGES_REQUESTED** [review]
It seems that we are switching the json library already: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33785

I suggest we wait until the switch is resolved and see what the possible implementation would look like.

**Comment by newly12** [comment]
updated code for the new json library, since the overhead of enabling `UseNumber` is high, I added the processing in a if else block, while the decode function when it is disabled remains the same as previous.

```
BenchmarkProcess
BenchmarkProcess-10             	   99046	     12432 ns/op	    9875 B/op	     114 allocs/op
BenchmarkProcessUseNumber
BenchmarkProcessUseNumber-10    	   74754	     15924 ns/op	   16730 B/op	     158 allocs/op
```

**Comment by ChrsMark** [comment]
Thank's @newly12.

>  since the overhead of enabling UseNumber is high

Would that make sense to mention this in the docs so as users know that enabling this option comes with a price?

File: pkg/stanza/docs/operators/json_parser.md [hunk 5]
```diff
@@ -4,16 +4,17 @@ The `json_parser` operator parses the string-type field selected by `parse_from`
...
+| `parse_from` | `body`           | The [field](../types/field.md) from which the value will be parsed. |
+| `parse_to`   | `attributes`     | The [field](../types/field.md) to which the value will be parsed. |
+| `on_error`   | `send`           | The behavior of the operator if it encounters an error. See [on_error](../types/on_error.md). |
+| `if`         |                  | An [expression](../types/expression.md) that, when set, will be evaluated to determine whether this operator should be used for the given entry. This allows you to do easy conditional parsing without branching logic with routers. |
+| `timestamp`  | `nil`            | An optional [timestamp](../types/timestamp.md) block which will parse a timestamp field before passing the entry to the output operator. |
+| `severity`   | `nil`            | An optional [severity](../types/severity.md) block which will parse a severity field before passing the entry to the output operator. |
+| `use_number` | `false`          | Numbers like `int` and `float` are parsed as `float64` by default, when `use_number` is enabled, numbers are parsed as `json.Number` and then converted to `int64` or `float64` based on the value. |
```
Comment (tiffany76) [inline pkg/stanza/docs/operators/json_parser.md:17]:
```suggestion
| `use_number` | `false`          | Numbers like `int` and `float` are parsed as `float64` by default. When `use_number` is enabled, numbers are parsed as `json.Number` and then converted to `int64` or `float64` based on the value. |
```

**Review by tiffany76 — COMMENTED** [review]
Just a small documentation suggestion. Thanks.

**Comment by newly12** [comment]
> Thank's @newly12.
> 
> > since the overhead of enabling UseNumber is high
> 
> Would that make sense to mention this in the docs so as users know that enabling this option comes with a price?

Done, PTAL.

**Review by ChrsMark — APPROVED** [review]
Looks good, thank's! Not sure if we could avoid this performance overhead, but since that's behind the specific setting I think we can land it and evaluate its adoption. 

@djaglowski please take a look as well.

**Review by djaglowski — APPROVED** [review]
This looks good to me, with one open question:

Should we use a different name for the parameter than `use_number`? To me the name seems to communicate an implementation detail, rather than the effect that it has. I would think something like `parse_ints` might be more informative. That said, I think the documentation is very clear so I don't think this is required.

**Comment by newly12** [comment]
> Looks good, thank's! Not sure if we could avoid this performance overhead, but since that's behind the specific setting I think we can land it and evaluate its adoption.
> 
> @djaglowski please take a look as well.

I checked fastjson code a little bit but looks like not much can be done apart from current approach, given limited options/interface exposed by the library.

**Comment by newly12** [comment]
> This looks good to me, with one open question:
> 
> Should we use a different name for the parameter than `use_number`? To me the name seems to communicate an implementation detail, rather than the effect that it has. I would think something like `parse_ints` might be more informative. That said, I think the documentation is very clear so I don't think this is required.

make sense to me, it is more informative, I've changed the config name to `parse_ints`, please take another look.

**Review by ChrsMark — APPROVED** [review]
LGTM. Just make sure to update the PR's title as well ;)

**Review by djaglowski — APPROVED** [review]

**Comment by djaglowski** [comment]
Thanks @newly12

**Pull request merged**